### ISSUE TITLE
[DX] [DependencyInjection] `Cake\Core\ServiceProvider` should throw exception when `$provides` is empty

### ISSUE THREAD

**Issue by Deltachaos** [cakephp/cakephp#17559]
### Description

The [documentation](https://book.cakephp.org/4/en/development/dependency-injection.html#service-providers) clearly states `Additionally those services must be defined in the $provides property.`.

But working a lot with other frameworks recently, from a DX perspecive it is very counterintuitive to not be confronted with an exception on invalid usage of a `ServiceProvider`.

My suggestion would be to thow a `LogicException` in `Cake\Core\ServiceProvider::provides` whenever `$this->provides` is empty.

### CakePHP Version

4.4, 5.0

**Comment by jamisonbryant** [comment]
I agree that the DX around using the service container and Cake could be better, especially when handling errors or unexpected conditions. For example, to someone new to using DI, what does “the service container lied about providing service X” even mean? I scratched my head about that one for a bit before I finally got enough experience to understand what was going on.

I think throwing exceptions with user-friendly messages as the OP states is a good idea for Cake’s DI feature set, especially in areas where footguns are commonly encountered such as service providers.
